README.md
# :books: Andre's Goodreads Toolbox, v1.25.1

![Maintenance](https://img.shields.io/maintenance/no/2022.svg)

9 Perl-scripts for Goodreads.com—the world largest book (cataloging) community. [What's new?](CHANGELOG.md)


## [recentrated.pl](./help/recentrated.md)

Checks all the books in your shelf for new ratings and notifies you
via periodical e-mail. It helps discover new criticisms and users with interesting 
libraries. You can [try this online](https://andre-st.github.io/goodreads/) if you 
don't want to install anything.
It implements the "follow book" feature that was requested in the Goodreads forums. 
[Usage+Screenshot](./help/recentrated.md)


## [friendrated.pl](./help/friendrated.md)

Prints all books rated 4 or 5 stars by 3 or more persons you follow (including friends). 
It implements the "books common among the people I follow" feature that was requested 
in the Goodreads forums. It also lists the most read authors, the most wished-for 
and hated books.
[Usage+Screenshot](./help/friendrated.md)


## [friendnet.pl](./help/friendnet.md)

Spiders your social network and creates files with edges and nodes which can be
easily processed with social network analysis software. It answers questions
like: Which members are popular among your friends?
[Usage+Screenshot](./help/friendnet.md)


## [friendgroup.pl](./help/friendgroup.md)

Prints discussion groups common among the persons you follow (including friends).
Searching groups on Goodreads is a PITA, and sometimes you don't know what you can have 
and wouldn't search for it. [Usage+Screenshot](./help/friendgroup.md)


## [likeminded.pl](./help/likeminded.md)

Prints Goodreads members who are interested in the same books as you.
It implements the "Finding people based on the books they've read"
feature that was requested in the Goodreads forums.
[Usage+Screenshot](./help/likeminded.md)


## [similarauth.pl](./help/similarauth.md)

Prints authors who Goodreads thinks are similar to all the authors you're reading.
It implements the "Finding [all] similar authors" feature that was requested in the 
Goodreads forums.
[Usage+Screenshot](./help/similarauth.md)


## [search.pl](./help/search.md)

Prints a books search result, ordered by average rating and number of ratings 
(most popular books), or date published, optionally with exact title matches. 
The Goodreads website doesn't offer it for some reason.
It implements the "Sort search results by rating" feature that was requested 
in the Goodreads forums.
[Usage+Screenshot](./help/search.md)


## [savreviews.pl](./help/savreviews.md)

Saves text-reviews for a book to a text-file. It implements the "Extract all 
reviews for a specific book" feature that was requested in the Goodreads forums.
[Usage+Screenshot](./help/savreviews.md)


## ~~[amz-tradein.pl](./help/amz-tradein.md)~~

This script fetched Amazon Trade-In prices for all books in a Goodreads.com
shelf ("resales" or "donations"). It automated regular manual bid-checking for 
hundreds of books, discovering sales opportunities. Amazon stopped its buyback 
program in 2015.
[Usage+Screenshot](./help/amz-tradein.md)



## Getting started

1a\.  [Docker](https://opensource.com/resources/what-docker) users can run the Toolbox in its own 
   container([?](https://www.docker.com/resources/what-container)),
   and view the results via web-browser at _localhost:8080_:

```console
$ docker run -it --publish=8080:80 ghcr.io/andre-st/goodreads-toolbox
```

1b\.  users without Docker can try to install the Toolbox directly on their systems:

```console
$ git clone https://github.com/andre-st/goodreads-toolbox.git
$ cd goodreads-toolbox
$ sudo make          # Gets required Perl modules from CPAN
```

2\.  at the prompt, try out the Toolbox programs:

```console
$ ./example-script.pl --help
```

Before [Docker for Windows or Mac](https://github.com/docker/toolbox/releases) 
and the project's Docker-images became available,
a Windows user wrote me that he ran the Toolbox on the [Windows 10 Subsystem for Linux](https://linuxhint.com/install_ubuntu_windows_10_wsl/) (WSL).


Long program runtimes: Goodreads slows down all requests and we have to load a lot of data.
  Start one program and do other things in the meantime.
  You can break any program with <kbd>CTRL</kbd>-<kbd>C</kbd> and continue later (reloads from a file-cache).



## Contributing

- Reporting bugs / feature requests
  - add a new issue via [Github's issue tracker](https://github.com/andre-st/goodreads-toolbox/issues/new)
  - [alternative contact options](AUTHORS.md)
  - thank you all who wrote me mails in the past or otherwise reported bugs and ideas :thumbsup:
- Writing your own scripts
  - see the [tests directory](./t/) for examples on how to use the toolbox library
  - see the [toolbox library documentation](./lib/Goodscrapes.pod)
  - [non-functional considerations](./t/README.md)
  - the [less complex issues](https://github.com/andre-st/goodreads-toolbox/labels/freshmen)
    would be good first issues to work on for users who want to contribute to this project



## Further readings

- About Goodreads
  - [GR developers group](https://www.goodreads.com/group/show/8095-goodreads-developers)
  - [GR technology stack](https://www.goodreads.com/jobs?id=597248#openPositions) 
		or [here](https://www.glasswaves.co/selected_projects.txt) 
		or [here](https://builtwith.com/goodreads.com) 
		or [DynamoDB+S3+Athena](https://aws.amazon.com/blogs/big-data/how-goodreads-offloads-amazon-dynamodb-tables-to-amazon-s3-and-queries-them-using-amazon-athena/)
  - [GR workplace reviews](https://www.glassdoor.com/Reviews/Goodreads-Reviews-E684833.htm), 
		anonymously about being acquired by Amazon, bureaucracy etc.
  - [GR on Crunchbase](https://www.crunchbase.com/organization/goodreads), 
		people, recent news & activity 
  - [GR members stats](https://www.statista.com/search/?q=goodreads&qKat=search) 
		or [here](https://qz.com/1106341/most-women-reading-self-help-books-are-getting-advice-from-men/) 
		or [here](https://onlinelibrary.wiley.com/doi/abs/10.1002/asi.23733)+[Sci-Hub](https://twitter.com/scihub_love) 
		or [here](https://book.pressbooks.com/chapter/goodreads-otis-chandler) 
		or [here](https://www.buzzfeednews.com/article/annanorth/what-amazon-is-getting-from-goodreads),
		source probably [Goodreads](https://www.goodreads.com/about/us)
  - [GR on Slideshare](https://www.slideshare.net/GoodreadsPresentations/presentations), 
		presenting GR book marketing to authors, see also [Author Feedback Group](https://www.goodreads.com/group/show/31471) 
  - [GR subreddit](https://www.reddit.com/r/goodreads/)
- Further software 
  - I leave statistics about your own reading habits to the following tools; 
		my toolbox, in contrast, focuses on the social periphery, with Goodreads providing the largest user base
  - Paul Klinger's [Bookstats](https://github.com/PaulKlinger/Bookstats) or [here](https://almoturg.com/bookstats/)
  - untested: John Smith's [GoodreadsAnalysis](https://github.com/JohnSmithDev/GoodreadsAnalysis/blob/master/REPORTS.md)
  - untested: Petr's [CompareBooks](https://github.com/vatioz/GoodreadsUserCompare) 
		browser [extension](https://chrome.google.com/webstore/detail/goodreads-compare-books/jcbnjaifalpejkcgfbpjbcmkfdildgpi) 
		adds "compare" info next to usernames
  - untested: Andrea Samorini's [SamoGoodreadsUtility](https://github.com/asamorini/goodreads.utility) 
		adds language filters to GR 
  - untested: Danish Prakash's [goodreadsh](https://github.com/danishprakash/goodreadsh) 
		is a command line interface for Goodreads (off. API)
  - untested: [Greasyfork Browser-Scripts](https://greasyfork.org/en/scripts/by-site/goodreads.com)
  - untested: the [Bookar Android app](https://github.com/intmainreturn00/Bookar) visualizes your books in augmented reality
  - untested: save your shelves and reviews [Goodreads data to SQLite](https://github.com/rixx/goodreads-to-sqlite)
  - Amazon: [export and filter long wishlists](https://github.com/andre-st/amazon-wishless) by priority and price (bargains)
- Other
  - Data: thousands of books and authors (not GR) https://openlibrary.org/developers/dumps
- Personal
  - [a list of things](./help/GOODTIPS.md) that improved my Goodreads experience: settings, browser extensions etc.
  - [Andre at Goodreads](https://www.goodreads.com/user/show/18418712-andr)




CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

Version number 1.MINOR.PATCH, increments:
- MINOR version when relevant functionality was added, removed, changed
- PATCH version on bug fixes, small aesthetic changes, config changes


## [Unreleased]

- see [GitHub issues labeled "next release"](https://github.com/andre-st/goodreads-toolbox/issues?q=is%3Aopen+is%3Aissue+label%3A"next+release")



## [v1.25.1] - 2023-01-29

### Fixed

- `greadbook()` works with the new (and old) Goodreads book page, now
- some unit test code was broken



## [v1.25] - 2022-09-22

After months, a necessary maintenance update without new features.

### Fixed

- many tools were broken due to the new Goodreads OpenID login procedure
  ([#38](https://github.com/andre-st/goodreads-toolbox/issues/38),
   [#39](https://github.com/andre-st/goodreads-toolbox/issues/39)),
  login (`glogin()`) should work now, unit tests are passing again
- "Warning: next input will be in plaintext" 
   on password prompt on some systems without Term::ReadKey module.
   If important to you, run `make deps`. 
- some unit test code was broken 


### Removed

- `cron.monthly` wasn't used for a long time so I don't want to maintain it any longer



## [v1.24] - 2022-03-10

This release does not add any new features. 
However, some major internal renewals took place.

**Note:** This will break your existing Toolbox installation.
Users have to run `make deps` and `make installdirs` after updating.

### Changed

- Perl5 module dependencies are not installed system-wide anymore.
	Now these go into the project's _local_ lib-directory.
	This helps non-Docker users to remove the toolbox with less residue.
	Also easens installation without root-privileges, 
	given that Perl's `local::lib` module is installed.

- replaced outdated Perl5 module `WWW::Curl::Easy` with `HTTP::Tiny`.
	This should fix Curl-based installation problems on newer systems;

- recentrated.pl:
	- no longer saves to `/var/db/good` but to `./list-out/recentrated/`
	- log-file written to `./list-out/recentrated/log.txt` now

- Docker:
	- image is based on Alpine Linux now
	- webserver webfs was replaced with thttpd



## [v1.23.3] - 2022-01-21

### Fixed

- login didn't worked anymore due to HTML changes on a Goodreads webpage
- `user_rating` bug when extracting own(?) shelf; shelves of other users still work with the old code

### Changed

- updated unit tests

### Added

- new `greadcomments()` function in the Toolbox library to extract a user's comments on reviews, status updates etc.



## [v1.23.1] - 2021-07-01

### Changed

- migrated from DockerHub to GitHub Packages (ghcr.io); DockerHub no longer supports GitHub integration for free (for new accounts)
- a larger review text is indicated by at least 500 characters now (was 160 characters)

### Added

- Makefile offers a Github Packages target
- clearer error message in a CAPTCHA situation



## [v1.23] - 2020-10-31

This is primarily a maintenance update after a long time.

### Fixed

- similaraut.pl:
	- failed to load authors because the source markup changed after the last version
- some improvements on the looks of all HTML reports

### Added

- similarauth.pl:
	- new `--minseen` parameter to drop authors that are not similar to at least N other authors;
	  reduces the number of authors and size of the output file

### Changed

- recentrated.pl:
	- `--textonly` option also drops smaller texts now,
	   because the flag's purpose is to cut down "noisy" e-mails with too many ratings
	   and smaller reviews are almost always useless too ("loved it so much!").
	   Without this flag, [\*\*\*\* ], [TTTT ] and [tttt ] are shown as usual.
- lib/Goodscrapes.pl:
	- greadreviews() parameter `text_only` was renamed to `text_minlen`



## [v1.22] - 2019-11-16
### Fixed

- possible cross-site scripting attacks (XSS) against our HTML reports
  by adding Javascript to reviews, usernames, book titles on the Goodreads website.
  This software is too insignificant to be a real target, but you never know.
  
### Added

- `Dockerfile`: Docker is a popular software that allows users to run apps
  in an isolated container with all dependencies included/matched; 
  there are make-targets for Docker, now, see `make help`.
  The container runs a simple web-server so that a host can access 
  any HTML report generated within the container.
  ([#30](https://github.com/andre-st/goodreads-toolbox/issues/30))
  

### Changed

- all programs write to the new `list-out` directory by default and 
  not to the main directory any longer.
  This simplifies sharing files between a Docker container and its host 
  (using Docker volumes or a httpd).
- the `dict` directory was renamed to `list-in`, 
  `dict/default.lst` was renamed to `list-in/dict.lst`



## [v1.21.1] - 2019-10-13
### Fixed

- savreviews.pl:
	- "Undefined subroutine &Goodscrapes::max called at Goodscrapes.pm"



## [v1.21.0] - 2019-10-10
### Fixed

- "Use of uninitialized value $uid in concatenation (.) or string at Goodscrapes.pm" 
  was caused by "NOT A BOOK" books, which actually have different book IDs but lack 
  author info etc

### Added

- all programs support the `--ignore-errors` option which disables retries and let
  a program keep going on despite of errors. This is useful when Goodreads has
  a bad day with permanent timeouts or other problems which often only affect
  a few resources while most of them can be obtained without problems.
  You also can re-run a program without this option, so that missing resources
  are loaded from the web again (and everything else from your local cache)

### Removed

- developers: function `gsetcache()` was removed in favor of `gsetopt( cache_days => int )`



## [v1.20.2] - 2019-08-31
### Fixed

- friendnet.pl:
	- missing ");" at the end (syntax error)
- friendrated.pl: 
	- poor table-caption in authors-report file

### Added


likeminded.pl
#!/usr/bin/env perl

#<--------------------------------- MAN PAGE --------------------------------->|

=pod

=head1 NAME

likeminded - finding people on Goodreads.com based on the books they've read


=head1 SYNOPSIS

B<likeminded.pl> 
[B<-m> F<number>] 
[B<-a> F<number>] 
[B<-x> F<number>] 
[B<-d> F<filename>] 
[B<-u> F<number>] 
[B<-c> F<numdays>] 
[B<-o> F<filename>] 
[B<-s> F<shelfname> ...] 
[B<-i>]
F<goodloginmail> [F<goodloginpass>]


=head1 OPTIONS

Mandatory arguments to long options are mandatory for short options too.

=over 4

=item B<-m, --common>=F<number>

value between 0 and 100. Members with 100% commonality have read *all* the
authors you did, which is unlikely, so better use lower values. 
Default is 5, that is, members who have read at least 5% of your authors.
There's a huge bulge of members with low commonality and just a few with 
higher commonality. Cut away the huge bulge, and check the rest manually.
In my tests, 5% cut away 99% of members. 
If you primarily read fiction and popular authors you should try a value
between 20 and 30. You can correct to a lower value later and restart
the program, it will run faster as it loads previous data from the cache.


=item B<-a, --maxauthorbooks>=F<number>

some authors list over 2000 books, either due to bad cataloging on the 
Goodreads site or too different editions which couldn't be combined.
Chewing them all would significantly increase the program's runtime.
Better we limit the number to the most N popular; default is 100


=item B<-x, --rigor>=F<numlevel>

we need to find members who rate the books of our authors, 
though Goodreads just shows a few ratings. 
We exploit ratings filters and the reviews-search to find more members:

 level 1 = filters-based search of book-raters (max 5400 ratings) - default
 level 2 = like 1 plus dict-search if >3000 ratings with stall-time of 2min
 level n = like 1 plus dict-search with stall-time of n minutes

Rigor level 0 is useless here (latest readers only), 
and 2+ (dict-search) has a bad cost/benefit ratio given hundreds of books.


=item B<-d, --dict>=F<filename>

default see section FILES


=item B<-u, --userid>=F<number>

check another member instead of the one identified by the login-mail 
and password arguments. You find the ID by looking at the shelf URLs.


=item B<-c, --cache>=F<numdays>

number of days to store and reuse downloaded data in F</tmp/FileCache/>,
default is 31 days. This helps with cheap recovery on a crash, power blackout 
or pause, and when experimenting with parameters. Loading data from Goodreads
is a very time consuming process.


=item B<-o, --outfile>=F<filename>

name of the HTML file where we write results to, 
default see section FILES


=item B<-s, --shelf>=F<shelfname>

name of the shelf with a selection of books, default is "#ALL#". 
If the name contains special characters use an URL-encoded name.
You can use this parameter multiple times if there is more than 1 shelf to
include (boolean OR operation), see the examples section of this man page.
Use B<--shelf>=shelf1,shelf2,shelf3 to intersect shelves (Intersection
requires password).


=item B<-i, --ignore-errors>

Don't retry on errors, just keep going. 
Sometimes useful if a single Goodreads resource hangs over long periods 
and you're okay with some values missing in your result.
This option is not recommended when you run the program unattended.




=item B<-?, --help>

show full man page

=back


=head1 FILES

F<./list-in/dict.lst>

F<./list-out/likeminded-$USERID-$SHELF.html>

F</tmp/FileCache/>


=head1 EXAMPLES

$ ./likeminded.pl login@gmail.com MyPASSword

$ ./likeminded.pl --shelf=science --shelf=music  login@gmail.com

$ ./likeminded.pl --shelf=animals,fiction login@gmail.com

$ ./likeminded.pl --outfile=./sub/myfile.html  login@gmail.com

$ ./likeminded.pl -c 31 -s read -m 5 -o myfile.html  login@gmail.com


=head1 REPORTING BUGS

Report bugs to <datakadabra@gmail.com> or use Github's issue tracker
L<https://github.com/andre-st/goodreads-toolbox/issues>


=head1 COPYRIGHT

This is free software. You may redistribute copies of it under the terms of
the GNU General Public License L<https://www.gnu.org/licenses/gpl.html>.
There is NO WARRANTY, to the extent permitted by law.


=head1 SEE ALSO

More info in ./help/likeminded.md


=head1 VERSION

2022-03-10 (Since 2018-06-22)

=cut

#<--------------------------------- 79 chars --------------------------------->|


use strict;
use warnings qw(all);
use locale;
use 5.18.0;

# Perl core:
use FindBin;
use local::lib "$FindBin::Bin/lib/local/";
use        lib "$FindBin::Bin/lib/";
use Time::HiRes   qw( time tv_interval );
use POSIX         qw( strftime floor locale_h );
use File::Spec; # Platform indep. directory separator
use IO::File;
use Getopt::Long;
use Pod::Usage;
# Third party:
# Ours:
use Goodscrapes;



# ----------------------------------------------------------------------------
# Program configuration:
# 
setlocale( LC_CTYPE, "en_US" );  # GR dates all en_US
STDOUT->autoflush( 1 );
gsetopt( cache_days => 31 );

our $TSTART     = time();
our $MINCOMMON  = 5;
our $MAXAUBOOKS = 100;
our $RIGOR      = 1;
our $DICTPATH   = File::Spec->catfile( $FindBin::Bin, 'list-in', 'dict.lst' );
our $OUTPATH;
our @SHELVES;
our $USERID;

GetOptions( 'common|m=i'         => \$MINCOMMON,
            'maxauthorbooks|a=i' => \$MAXAUBOOKS,
            'rigor|x=i'          => \$RIGOR,
            'dict|d=s'           => \$DICTPATH,
            'userid|u=s'         => \$USERID,
            'outfile|o=s'        => \$OUTPATH,
            'shelf|s=s'          => \@SHELVES,
            'ignore-errors|i'    => sub{  gsetopt( ignore_errors => 1 );   },
            'cache|c=i'          => sub{  gsetopt( cache_days => $_[1] );  },
            'help|?'             => sub{  pod2usage( -verbose => 2 );      }) 
	or pod2usage( 1 );

pod2usage( 1 ) if !$ARGV[0];

glogin( usermail => $ARGV[0],  # Login also allows to load 200 books in 1 request
        userpass => $ARGV[1],  # Asks pw if omitted
        r_userid => \$USERID );

@SHELVES = qw( %23ALL%23 )
	if !@SHELVES;
	
$OUTPATH = File::Spec->catfile( $FindBin::Bin, 'list-out', sprintf( "likeminded-%s-%s.html", $USERID, join( '-', @SHELVES )))
	if !$OUTPATH;



# ----------------------------------------------------------------------------
# Primary data structures:
# 
my %authors;          # {$auid   => %author}
my %authors_read_by;  # {$userid}->{$auid => 1}
my %readers;          # {$userid => %user}
my %books;            # {$bookid => %book}, just check 1 book if 2 authors



# ----------------------------------------------------------------------------
# Query authors from the user's shelves:
#
printf( "Loading authors from \"%s\"...", join( '" and "', @SHELVES ));

greadauthors( from_user_id    => $USERID, 
              ra_from_shelves => \@SHELVES,
              rh_into         => \%authors, 
              on_progress     => gmeter( 'authors' ));



# ----------------------------------------------------------------------------
# Query all books of the loaded authors:
# 
my $audone  = 0;
my $aucount = scalar keys %authors;

die( $GOOD_ERRMSG_NOBOOKS ) unless $aucount;

printf( "\nLoading books of %d authors:\n", $aucount );

for my $auid (keys %authors)
{
	my $t0 = time();
	printf( "[%3d%%] %-25s #%-8s\t", 
			++$audone/$aucount*100, 
			$authors{$auid}->{name}, 
			$auid );
	
	my $imgurlupdatefn = sub{ $authors{$auid} = $_[0]->{rh_author} };  # TODO ugly
	
	greadauthorbk( author_id   => $auid,
	               limit       => $MAXAUBOOKS,
	               rh_into     => \%books,
	               on_book     => $imgurlupdatefn,
	               on_progress => gmeter( 'books' ));
	
	printf( "\t%6.2fs\n", time()-$t0 );
}
say "Done.";



# ----------------------------------------------------------------------------
# Query readers of all author books:
# Lot of duplicates (not combined as editions), but with unique reviewers tho
# 
my $bocount = scalar keys %books;
my $bodone  = 0;

printf( "Loading readers of %d author books:\n", $bocount );

for my $b (values %books)
{
	printf( "[%3d%%] %-40s  #%-8s\t", 
			++$bodone/$bocount*100, 
			substr( $b->{title}, 0, 40 ), 
			$b->{id} );

	my $t0 = time();
	my %revs;

	greadreviews( rh_for_book => $b, 
	              rh_into     => \%revs,
	              rigor       => $RIGOR,  
	              dict_path   => $DICTPATH,
	              on_progress => gmeter( 'memb' ));
	
	$authors_read_by{ $_->{rh_user}->{id} }{ $b->{rh_author}->{id} } = 1  # (2,3,5,5,3) $authors{id}->user_avg vs rev_user's avg author rating
		foreach( values %revs );
	
	printf( "\t%6.2fs\n", time()-$t0 );
}
say "Done.";



# ----------------------------------------------------------------------------
# Query additional info on all readers with P percent common books.
# We need to know their library sizes to calc a similarity ranking ("match") 
# (Github #18). At the same time, we can get profile pics for the report.
# 
# Note: The match-value isn't coverage and percentage but a relation of the
# number of common authors to the number of all books in the match-mate's
# library - it's not a as good as comparing #authors to #authors (coverage)
# but it comes close. Comparing #authors would require loading all the books
# of the match-mates which takes forever as opposed to a single request
# for the number of books. Think of the match-value as stars or points
# or score, not percent - the higher the better.
#
# 
printf( "Dropping who read less than %d%% of your authors... ", $MINCOMMON );

my $bycount0 = scalar keys %authors_read_by;

for my $userid (keys %authors_read_by)
{
	my $aucommon     = scalar keys %{$authors_read_by{$userid}};
	my $aucommonperc = int($aucommon/$aucount*100+0.5);  # Rounded
	
	delete $authors_read_by{$userid}
		if $aucommonperc < $MINCOMMON || $userid eq $USERID;  # Drops ~99%
}

my $bycount1 = scalar keys %authors_read_by;

printf( "-%d memb (%3.3fs%%)\n", 
		$bycount0-$bycount1, 
		100-($bycount1/$bycount0*100) );



my $ucount = scalar keys %authors_read_by;
my $udone  = 0;

printf( "Loading profiles of the remaining %d members:\n", $ucount );

for my $userid (keys %authors_read_by)
{
	printf( "[%3d%%] goodreads.com/user/show/%-8s", 
			++$udone/$ucount*100, 
			$userid );
	
	my $t0 = time();
	my %u  = greaduser( $userid );
	
	next unless %u;  # Probably due to errors (Timeouts etc)
	
	printf( "\t%6.2fs", time()-$t0 );
	
	print ( "\tprivate account\n" ) and next 
		if $u{num_books} == 0 || $u{is_private};
	
	
	$u{ aucommon }    = scalar keys %{$authors_read_by{$userid}};	
	$u{ match    }    = int( $u{aucommon}/$u{num_books}*1000 + 0.5 ); # Watch div by zero!
	$readers{$userid} = \%u;
	
	print( "\t".( "*" x ($u{match}/10) )."\n" );
}

say "Done.";



# ----------------------------------------------------------------------------
# Write results to HTML file:
# 
printf( "Writing report (N=%d) to \"%s\"... ", scalar keys %readers, $OUTPATH );

my $fh  = IO::File->new( $OUTPATH, 'w' ) or die "[FATAL] Cannot write to $OUTPATH ($!)";
my $now = strftime( '%a %b %e %H:%M:%S %Y', localtime );
my $shv = sprintf( "%s <q>%s</q>", 
                   (scalar @SHELVES > 1 ? 'shelves' : 'shelf'), 
                   join( '</q> and <q>', @SHELVES ) );

print $fh ghtmlhead( "Members who read at least ${MINCOMMON}% of the authors in ${USERID}'s ${shv}, on $now",
		[ '<Rank:', 'Match', '!Common Authors' ]);

my $line;
for my $userid (sort{ $readers{$b}->{match} <=> 
                      $readers{$a}->{match} } keys %readers)
{
	$line++;
	print $fh qq{
			<tr>
			<td>$line</td>
			<td>
			  <a href="https://www.goodreads.com/user/show/${userid}" target="_blank">
			    <img src="${\ghtmlsafe( $readers{$userid}->{img_url} )}"
			             >${\ghtmlsafe( $readers{$userid}->{name}    )}
			    <br>
			    <small>
			      ${\ghtmlsafe( $readers{$userid}->{aucommon}  )}&nbsp;authors&nbsp;over<br>
			      ${\ghtmlsafe( $readers{$userid}->{num_books} )}&nbsp;books
			    </small>
			  </a>
			</td>
			<td>
			};
			
	print $fh qq{
			<gr-author><img src="${\ghtmlsafe( $authors{$_}->{img_url} )}"
			                    >${\ghtmlsafe( $authors{$_}->{name}    )}</gr-author>
			} foreach (keys %{$authors_read_by{$userid}});

	print $fh qq{
			</td>
			</tr>
			};
}

print $fh ghtmlfoot();
undef $fh;



# ----------------------------------------------------------------------------
# Done:
# 
printf( "\nTotal time: %.0f minutes\n", (time()-$TSTART)/60 );


